src/gui/src/SongEditor/SongEditor.cpp

```diff
@@ -811,6 +813,28 @@ void SongEditor::paintEvent( QPaintEvent *ev )
 	QPainter painter(this);
 	painter.drawPixmap( ev->rect(), *m_pSequencePixmap, ev->rect() );
 
+	// Draw the song position cursor
+	if ( pHydrogen->getSong()->getMode() != Song::PATTERN_MODE ) {
+		float fPos = pHydrogen->getPatternPos();
+		if ( pHydrogen->getCurrentPatternList()->size() != 0 ) {
+			int nLength = pHydrogen->getCurrentPatternList()->longest_pattern_length();
+			fPos += (float)pHydrogen->getTickPosition() / (float)nLength;
+		}
+		else {
+			// nessun pattern, uso la grandezza di default
+			fPos += (float)pHydrogen->getTickPosition() / (float)MAX_NOTES;
+		}
+		uint x = (int)( m_nMargin + fPos * m_nGridWidth - 11 / 2 );
```

Comment:
818-827 here are duplicated from the ruler's paintEvent. Can it be factored out? Possibly by moving to the SongEditorPanel? Or even a static function as it doesn't look like it needs anything from the SongEditor or SongEditorRuler.

Reply:
There actually is a dependence in line 827 on the two members `m_nMargin` and `m_nGridWidth`, whose values are duplicated between the `SongEditor`, `SongEditorPositionRuler`, and in the case of the grid width, `AutomationPathView`.  So maybe that block should be a member of the SongEditorPanel, and those members moved as well?

Reply:
> There actually is a dependence in line 827 on the two members `m_nMargin` and `m_nGridWidth`, whose values are duplicated between the `SongEditor`, `SongEditorPositionRuler`, and in the case of the grid width, `AutomationPathView`. So maybe that block should be a member of the SongEditorPanel, and those members moved as well?

Ah, sorry I didn't spot those! Yes. What *really* wants to happen here is that there should be a common ancestor class of AutomationPath, SongEditor, SongEditorRuler etc. but failing that, cut-and-paste is fine (this is probably a controversial opinion!) because it means the code that should be commonised is easier to spot and easier to unify when such a class exists :)